tests/Custom/WebDriverTest.php

```diff
@@ -2,32 +2,94 @@
 
 namespace Mink\WebdriverClassicDriver\Tests\Custom;
 
-use Behat\Mink\Tests\Driver\TestCase;
+use Behat\Mink\Exception\DriverException;
+use Mink\WebdriverClassicDriver\Tests\WebdriverClassicConfig;
 use Mink\WebdriverClassicDriver\WebdriverClassicDriver;
+use PHPUnit\Framework\TestCase;
 
 class WebDriverTest extends TestCase
 {
+    private WebdriverClassicDriver $driver;
+
     protected function setUp(): void
     {
         parent::setUp();
 
-        $this->getSession()->start();
+        $this->driver = WebdriverClassicConfig::getInstance()->createDriver();
     }
 
     protected function tearDown(): void
     {
-        $this->getSession()->stop();
+        if ($this->driver->isStarted()) {
+            $this->driver->stop();
+        }
 
         parent::tearDown();
     }
 
-    public function testGetWebDriverSessionId(): void
+    public function testDriverMustBeStartedBeforeUse(): void
+    {
+        $this->expectException(DriverException::class);
+        $this->expectExceptionMessage('Base driver has not been created');
+
+        $this->driver->visit('https://example.com');
+    }
+
+    public function testDriverMustBeStartedBeforeItCanBeStopped(): void
+    {
+        $this->expectException(DriverException::class);
+        $this->expectExceptionMessage('Driver has not been started');
+
+        $this->driver->stop();
+    }
+
+    public function testStartedDriverCannotBeSubsequentlyStarted(): void
     {
-        $driver = $this->getSession()->getDriver();
-        assert($driver instanceof WebdriverClassicDriver);
-        $this->assertNotEmpty($driver->getWebDriverSessionId(), 'Started session should have an ID');
+        $this->driver->start();
+
+        $this->expectException(DriverException::class);
+        $this->expectExceptionMessage('Base driver has already been created');
+
+        $this->driver->start();
+    }
 
-        $driver = new WebdriverClassicDriver();
-        $this->assertNull($driver->getWebDriverSessionId(), 'Non-started session should not have an ID');
+    public function testDriverCatchesUpstreamErrorsDuringStart(): void
+    {
+        $driver = $this->createPartialMock(WebdriverClassicDriver::class, ['createWebDriver', 'getWebDriver']);
+        $driver
+            ->expects($this->once())
+            ->method('createWebDriver')
+            ->willThrowException(new \RuntimeException('An upstream error'));
+
+        $this->expectException(DriverException::class);
+        $this->expectExceptionMessage('Could not start driver: An upstream error');
+
+        $driver->start();
+    }
+
+    public function testDriverCatchesUpstreamErrorsDuringStop(): void
+    {
+        $driver = $this->createPartialMock(WebdriverClassicDriver::class, ['createWebDriver', 'isStarted', 'getWebDriver']);
```

Comment:
Partial mocks usually indicate bad architectural decisions (that's why Prophecy library disallows them).

If we would be using some kind of factory, that internally called the `RemoteWebDriver::create` method in the `\Mink\WebdriverClassicDriver\WebdriverClassicDriver::start`, then we could have injected a new factory for testing, that returned a mock of the `RemoteWebDriver` class.

Maybe there are other architectural solutions to this.

Reply:
In this case it's fine, having a factory for the base driver feels like overengineering to me.

Reply:
OK.